src/rust/src/contract.rs

```diff
@@ -2,31 +2,106 @@
 // under the Apache License, Version 2.0. See the COPYING file at the root
 // of this distribution or at http://www.apache.org/licenses/LICENSE-2.0
 
-use crate::log::partition::TX;
+use crate::{log::partition::TX, rust_bridge::XDRBuf};
 use log::info;
 use std::io::Cursor;
 
-use cxx::{CxxString, CxxVector};
+use cxx::CxxString;
+use im_rc::OrdMap;
 use std::error::Error;
 use stellar_contract_env_host::{
-    xdr::{ReadXdr, ScVec, WriteXdr},
-    Host, VM,
+    storage, xdr,
+    xdr::{
+        ContractDataEntry, LedgerEntry, LedgerEntryData, ReadXdr, ScObject, ScStatic, ScVal, ScVec,
+        WriteXdr,
+    },
+    Host, HostError, Vm,
 };
 
 /// Deserialize an SCVec XDR object of SCVal arguments from the C++ side of the
 /// bridge, instantiate a Host and VM with the provided WASM, invoke the
 /// requested function in the WASM, and serialize an SCVal back into a return
 /// value.
 pub(crate) fn invoke_contract(
-    wasm: &CxxVector<u8>,
+    contract_id: &XDRBuf,
     func: &CxxString,
-    args: &CxxVector<u8>,
+    args: &XDRBuf,
+    footprint: &XDRBuf,
+    ledger_entries: &Vec<XDRBuf>,
 ) -> Result<Vec<u8>, Box<dyn Error>> {
-    let arg_scvals = ScVec::read_xdr(&mut Cursor::new(args.as_slice()))?;
+    let contract_id = xdr::Hash::read_xdr(&mut Cursor::new(contract_id.data.as_slice()))?;
```

Comment:
Nitpick: this function desperately needs whitespace, and probably some comments too.

Reply:
I'll do some factoring and commenting but I can't do much about the whitespace -- that's how rustfmt formats it!

Reply:
You should be able to add empty lines. I don't think rustfmt will remove them?

Reply:
Perhaps? Anyway I split it into several support functions, much clearer now.